src/components/Hoverable/ActiveHoverable.tsx

```diff
@@ -0,0 +1,143 @@
+import CONST from '@src/CONST';
+import HoverableProps from './types';
+
+/**
+ * Assigns a ref to an element, either by setting the current property of the ref object or by calling the ref function
+ * @param ref The ref object or function.
+ * @param element The element to assign the ref to.
+ */
+function assignRef(ref: ((instance: HTMLElement | null) => void) | MutableRefObject<HTMLElement | null>, element: HTMLElement) {
+    if (!ref) {
+        return;
+    }
+    if (typeof ref === 'function') {
+        ref(element);
+    } else if ('current' in ref) {
+        // eslint-disable-next-line no-param-reassign
+        ref.current = element;
+    }
+}
+
+type ActiveHoverableProps = Omit<HoverableProps, 'disabled'>;
+
+function ActiveHoverable({onHoverIn, onHoverOut, shouldHandleScroll, children}: ActiveHoverableProps, outerRef: Ref<HTMLElement>) {
+    const [isHovered, setIsHovered] = useState(false);
+
+    const ref = useRef<HTMLElement | null>(null);
+    const isScrolling = useRef(false);
+    const isHoveredRef = useRef(false);
+
+    const updateIsHovered = useCallback(
+        (hovered: boolean) => {
+            isHoveredRef.current = hovered;
+            if (shouldHandleScroll && isScrolling.current) {
+                return;
+            }
+            setIsHovered(hovered);
+        },
+        [shouldHandleScroll],
+    );
+
+    // Expose inner ref to parent through outerRef. This enable us to use ref both in parent and child.
+    useImperativeHandle<HTMLElement | null, HTMLElement | null>(outerRef, () => ref.current, []);
+
+    useEffect(() => (isHovered ? onHoverIn?.() : onHoverOut?.()), [isHovered, onHoverIn, onHoverOut]);
+
+    useEffect(() => {
+        if (!shouldHandleScroll) {
+            return;
+        }
+
+        const scrollingListener = DeviceEventEmitter.addListener(CONST.EVENTS.SCROLLING, (scrolling) => {
+            isScrolling.current = scrolling;
+            if (!isScrolling.current) {
+                setIsHovered(isHoveredRef.current);
+            }
+        });
+
+        return () => scrollingListener.remove();
+    }, [shouldHandleScroll]);
+
+    useEffect(() => {
+        // Do not mount a listener if the component is not hovered
+        if (!isHovered) {
+            return;
+        }
+        /**
+         * Checks the hover state of a component and updates it based on the event target.
+         * This is necessary to handle cases where the hover state might get stuck due to an unreliable mouseleave trigger,
+         * such as when an element is removed before the mouseleave event is triggered.
+         * @param event The hover event object.
+         */
+        const unsetHoveredIfOutside = (event: MouseEvent) => {
+            if (!ref.current || ref.current.contains(event.target as Node)) {
+                return;
+            }
+
+            setIsHovered(false);
+        };
+
+        document.addEventListener('mouseover', unsetHoveredIfOutside);
```

Comment:
We could extract the utility, but we need to mkae sure the entire functionality of `addEventListener` and `removeEventListener` is accessible from outside (e.g. listener's `options` argument). I will give it a moment after all the minor changes to the PR are resolved 👍

Reply:
Yes, you're right. Also, it would be great to hook into TypeScript's type mapping for HTML event types.

Reply:
After careful consideration, I found a few reasons not to implement such solution or at least to not do it here when the scope is quite different.

Here is a implementation that I made in order to test the suggestion:

```ts
const useEventListener = (element: HTMLElement, ...listenerParams: Parameters<typeof addEventListener>) => {
    useEffect(() => {
        element.addEventListener(...listenerParams);
        return () => element.removeEventListener(...listenerParams);
    }, [element, listenerParams]);
};
```

Here are my concerns:

- every listener (callback) needs to be memoized now, and we cannot simply inline it as there is no access to `useEffect` (see effect in line 77 of `ActiveHoverable`)
- cannot opt out of listening if not needed (see effect in line 53 of `ActiveHoverable`)

Let me know what you think about it 👍

Reply:
I think something like this would work:

```ts
function useTargetEventEffect<Ev>(
    enabled: boolean,
    target: EventTarget,
    type: string,
    listener: (ev: Ev) => void,
    options?: boolean | AddEventListenerOptions,
): void {
    useEffect(() => {
        if (!enabled) {
            return;
        }

        target.addEventListener(type, listener as EventListener, options);

        return () => target.removeEventListener(type, listener as EventListener, options);
    }, [enabled, target, type, listener, options]);
}

function useDocumentEventEffect<K extends keyof DocumentEventMap>(
    enabled: boolean,
    type: K,
    listener: (ev: DocumentEventMap[K]) => void,
    options?: boolean | AddEventListenerOptions,
): void {
    useTargetEventEffect(
        enabled,
        document,
        type,
        listener,
        options,
    );
}

function useElementEventEffect<K extends keyof HTMLElementEventMap>(
    enabled: boolean,
    element: HTMLElement,
    type: K,
    listener: (ev: HTMLElementEventMap[K]) => void,
    options?: boolean | AddEventListenerOptions,
): void {
    useTargetEventEffect(
        enabled,
        element,
        type,
        listener,
        options,
    );
}
```

But I'm not 100% sure, this would need to be carefully tested. We can consider this out of scope for now.